tests/test_constrain_current.py

```diff
@@ -16,17 +16,35 @@ class _ExactValueProfile:
     def __init__(self, eq, grid):
         self.eq = eq
         self.grid = grid
+        self.transforms = get_transforms("current_rr", eq=eq, grid=grid)
+        self.profiles = get_profiles("current_rr", eq=eq, grid=grid)
+        self.params = get_params("current_rr", eq=eq)
 
     def compute(self, params, dr, *args, **kwargs):
         if dr == 0:
             # returns the surface average of B_theta in amperes
-            return self.eq.compute("current", grid=self.grid)["current"]
```

Comment:
Good job catching this.

This test was failing on this PR before this last commit when `self.grid` was `ConcentricGrid`.  The assertion statement caught that iota on the 2 surfaces closest to the axis didn't match with the given power series.

 This was because the `profiles["current"].compute` call [here](https://github.com/PlasmaControl/DESC/blob/cb2313a7d03397a2f71904a64896da766643a2cc/desc/compute/_profiles.py#L556) was monkey patched to call this function. In this PR, `self.eq.compute("current", grid=self.grid)["current"]` now does a surface average on on a `LinearGrid` rather than the `self.grid` in the argument. So when computing `iota` in this test (before the last commit), we were doing a surface average on a `LinearGrid` as well as on `ConcentricGrid` [here](https://github.com/PlasmaControl/DESC/blob/cb2313a7d03397a2f71904a64896da766643a2cc/desc/compute/_profiles.py#L577).

Ideally this shouldn't matter because the surface average should yield the same result regardless of what grid is used.
But I think the fact that `ConcentricGrid` has few nodes on the surface closest to the axis was causing that discrepancy.

Reply:
@dpanici
@ddudt
Last week we had thought it was weird why a surface average on a linear grid and concentric grid were not exactly the same (not just different at the axis).
I think this is expected though.
The difference between them decreases with resolution increase, meaning the accuracy of the average being captured is dependent on how well midpoint rule integration approximates the true average, and the linear grid has more nodes.

```python
eq = get("W7-X")
L, M, N, NFP, sym = eq.L_grid, eq.M_grid, eq.N_grid, eq.NFP, eq.sym
cg = ConcentricGrid(L=L, M=M, N=N, NFP=NFP, sym=sym)
rho = np.copy(compress(cg, cg.nodes[:, 0]))
lg = LinearGrid(rho=rho, M=M, N=N, NFP=NFP, sym=sym)

def compute_iota(data, grid):
    current_term = (
        mu_0
        / (2 * np.pi)
        * data["current"]
        / data["psi_r"]
    )
    num = (
        data["lambda_z"] * data["g_tt"] - (1 + data["lambda_t"]) * data["g_tz"]
    ) / data["sqrt(g)"]
    den = data["g_tt"] / data["sqrt(g)"]
    num_avg = surface_averages(grid, num)
    den_avg = surface_averages(grid, den)
    return (current_term + num_avg) / den_avg

    
data_cg = eq.compute(["current", "psi_r", "lambda_z", "lambda_t", "g_tt", "g_tz", "sqrt(g)"], grid=cg)
data_lg = eq.compute(["current", "psi_r", "lambda_z", "lambda_t", "g_tt", "g_tz", "sqrt(g)"], grid=lg)

iota_cg = compress(cg, compute_iota(data_cg, cg))
iota_lg = compress(lg, compute_iota(data_lg, lg))
print("absolute difference")
print(abs(iota_cg - iota_lg))
print("relative difference")
print(abs(iota_cg - iota_lg) / abs(iota_lg))
```
```python
absolute difference
[4.99024289e-01 1.03736010e-03 1.00308820e-05 1.26244516e-06
 3.89644578e-07 8.73197681e-09 3.73933381e-08 2.29957960e-08
 5.23922110e-08 7.08192308e-08 3.95633758e-08 1.11696806e-08
 3.24810173e-08 1.07613550e-08 9.37128031e-09 1.99362553e-08
 0.00000000e+00]
relative difference
[5.82936911e-01 1.21172527e-03 1.17144310e-05 1.47355976e-06
 4.54329755e-07 1.01627698e-08 4.33882724e-08 2.65566947e-08
 6.00864063e-08 8.04441105e-08 4.43891265e-08 1.23499529e-08
 3.53475139e-08 1.15298979e-08 9.90430237e-09 2.08536925e-08
 0.00000000e+00]
```

Reply:
The concentric grid and linear grid have the same nodes out the outermost surface, so they should agree there, but as you go in the concentric grid looses some nodes so we expect them to differ, which lines up with the values you got. The variation of the function over the surface also decreases as you go in, so its not that big a difference (relative error < 0.1% throughout most of the volume)